make-derivation.nix: lib.warn if drv.__spliced missing

amjoseph-nixpkgs opened the pull request:
This should wait until after 23.11, but I'd like to try to merge it fairly promptly thereafter.

Right now when an unspliced package appears as a dependency, we just silently do the wrong thing, cross our fingers, and hope for the best.  This is the reason why troubleshooting cross compilation is such a headache.

This PR adds a `lib.warn` for this situation *when cross compiling*.  So we still do the wrong thing, but at least we tell the user that something is probably wrong.  The warning will also cause CI to fail.  I fixed about a dozen of these failures.  The `list-of-shame` provides an escape hatch for urgent situations.

The `list-of-shame` currently has (about) two dozen of these splicing failures; these are ignored in order to get CI to pass.  The idea is to get a check like this in place **to prevent even more of these bugs from occurring** while we work to eliminate the ones that are already there.  Since `lib.warn` causes CI failures this will prevent people from introducing any more of these bugs (or else force them to add their stinky deeds to the list of shame).

- [x] CI passes
- [x] No mass rebuild (14 packages rebuilt, due to fixing splice failures)

Commit:
stdenv: lib.warn if drv.__splice is missing when expected

This commit adds a `lib.warn` check to stdenv.mkDerivation, causing
it to emit a warning if a dependency which ought to be spliced in
fact is not.

amjoseph-nixpkgs commented:
Proof that we are not splicing Rust hooks: check out this PR, then run:

```
nix-instantiate . -A pkgsCross.aarch64-multiplatform.fd
```

in the output you'll see:

```
trace: warning: derivation fd:
                     unspliced dependency /nix/store/99jm62fd7xh1yzjys0z3r70d1vvra51r-cargo-build-hook.sh
                       build=x86_64-unknown-linux-gnu
                       target=aarch64-unknown-linux-gnu

trace: warning: derivation !!no pname!!:
                     unspliced dependency /nix/store/cxxbmvm5dsjif4v4d82q0mb7ympp8jlc-aarch64-unknown-linux-gnu-cargo-1.72.1
                       build=x86_64-unknown-linux-gnu
                       target=aarch64-unknown-linux-gnu

trace: warning: derivation fd:
                     unspliced dependency /nix/store/84aap5fr5s6yblga5w6fbx71j21l591a-cargo-check-hook.sh
                       build=x86_64-unknown-linux-gnu
                       target=aarch64-unknown-linux-gnu

trace: warning: derivation fd:
                     unspliced dependency /nix/store/0796vgahxgg8n7w1gr2pdb9sd3kibk13-cargo-install-hook.sh
                       build=x86_64-unknown-linux-gnu
                       target=aarch64-unknown-linux-gnu

trace: warning: derivation fd:
                     unspliced dependency /nix/store/qppx46087vy5xx8j3jajbzbqsjhvcp2b-cargo-setup-hook.sh
                       build=x86_64-unknown-linux-gnu
                       target=aarch64-unknown-linux-gnu
```

Artturin commented:
Is similar to https://github.com/NixOS/nixpkgs/pull/204387

amjoseph-nixpkgs commented:
> Is similar to #204387

Well, I think they're sort of orthogonal (and both are a good idea).  It looks like `checkExecutableOnBuild`  always succeeds when the `__spliced` attribute is missing even if it should be there -- whereas the point of this PR is to emit a warning in exactly that situation.  Currently it takes way too long to figure out if a problem is caused by splicing not happening.

You should merge yours first since mine isn't done and is easier to rebase.

Artturin commented:
Befofe merge there should be a way to fix the warning.

We should find out a way to splice or work around the lack of splicing when a function that returns derivation(s) is used (without adding it to be reachable from top-level).

Maybe a callPackage like func which internally calls with the various sets callPackage's, but a issue would still be what the user passes if it was not already spliced (eg overrideAttrs was used).


> always succeeds when the __spliced attribute is missing even if it should be there 

Explicitly used sets like buildPackages won't have a __spliced in packages

amjoseph-nixpkgs commented:
> We should find out a way to splice or work around the lack of splicing when a function that returns derivation(s) is used (without adding it to be reachable from top-level).

The way to do that is to have all derivations be spliced, always, all the time.  Then we don't have to worry about it.

Commit:
gcc: use stdenv.cc.bintools instead of targetPackages_bintools in buildInputs

Commit:
isl: fix splicing failure

Commit:
stdenv: allow unspliced hostTarget dependencies (for now)

Artturin commented on pkgs/development/compilers/gcc/common/dependencies.nix:
```diff
@@ -88,5 +85,8 @@ in
-  depsTargetTarget = optionals (!withoutTargetLibc && threadsCross != { } && threadsCross.package != null) [ threadsCross.package ];
+  depsTargetTarget = optionals (!withoutTargetLibc && threadsCross != { } && threadsCross.package != null) [ threadsCross.package ]
+  ++ [
+    stdenv.cc.bintools # For linking code at run-time
```
```
nix-repl> pkgsCross.aarch64-multiplatform.__splicedPackages.stdenv.cc.bintools ? __spliced
false
```

amjoseph-nixpkgs replied:
I think it's okay; nothing uses `stdenv.cc.bintools` as a dependency AFAICT -- it's mainly there "for reference purposes".

Put another way,

```
nix-instantiate . -A pkgsCross.aarch64-multiplatform.bash
```

emits no warnings.

amjoseph-nixpkgs replied:
Latest push adds a `targetPackages_bintools` to the top level, which is basically just `targetPackages.stdenv.cc.bintools` except that it gets spliced.

amjoseph-nixpkgs commented:
> Explicitly used sets like buildPackages won't have a __spliced in packages

I think that's the point.  Those should *never* be used in depsFooBar.

They should only be used when you need to interpolate into a string, like `"${buildPlatform.lndir} ..."`

Commit:
stdenv: add list-of-shame

Co-authored-by: Artturi <Artturin@artturin.com>

amjoseph-nixpkgs commented:
> Befofe merge there should be a way to fix the warning.

See 083501cc7f7b

Artturin commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9510,7 +9510,7 @@ with pkgs;
   ised = callPackage ../tools/misc/ised { };
 
-  isl = isl_0_20;
+  isl = __splicedPackages.isl_0_20;
```
This probably wasn't a problem because `isl` would be spliced again

```
nix-repl> pkgsCross.aarch64-multiplatform.__splicedPackages.isl ? __spliced
true
```

amjoseph-nixpkgs replied:
You're right; it's only the other `isl` that needs it.

I'll push a commit to drop this as soon as ofborg finishes the current eval, which I think might actually succeed!  Don't want to interrupt it just yet.

amjoseph-nixpkgs replied:
9c46b1a6517b

Artturin commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -1,3 +1,42 @@
+          "setuptools-check-hook"
+          "unittest-check-hook"
+
+          # shame on you, llvm!
+          "llvm"
+          "compiler-rt"
+          "libcxx"
+          "libcxxabi"
```
the llvm sets use the normal `makeScope` and that's why there's no `__spliced`'s  in the sets own `callPackage`

I'm little by little working on making the sets better https://github.com/NixOS/nixpkgs/issues/253671
once I dedupe more stuff it'll be much easier to dedupe the `llvm/*/default.nix`'es

Nice list name.

Artturin commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -1,3 +1,49 @@
+          "arm-none-eabi-stage-final-gcc-wrapper"
+
+          # other shameful derivations
+          "mcfgthreads"
```
```suggestion
          # has overrides which discard splicing
          # improvement: convert makeScope to makeScopeWithSplicing'
          # may not fix it but it'll improve the set
          # https://github.com/NixOS/nixpkgs/blob/53aa767c849b159cdb8c59dce4a5a44f167fc31b/pkgs/os-specific/windows/default.nix#L20
          "mcfgthreads"
```

Commit:
rustc-wasm32: give it a custom pname for stdenv list-of-shame

Commit:
cracklib: do not use buildPackages for self-reference (fails to splice)

amjoseph-nixpkgs commented:
Force-push has cleanup, dropped the stuff that I moved to other PRs.

I think 23888fe might actually pass CI.

amjoseph-nixpkgs commented:
Yeah it's gonna pass CI.

SuperSandro2000 commented on pkgs/development/libraries/cracklib/default.nix:
```diff
@@ -15,7 +16,7 @@ stdenv.mkDerivation rec {
   };
 
-  nativeBuildInputs = lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) buildPackages.cracklib;
+  nativeBuildInputs = lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) cracklib;
```
I am trying to get more fit in cross: Shouldn't this use canExecute?

amjoseph-nixpkgs replied:
If this is what you mean, then yes, that would be an improvement (let's put it in a separate PR though -- one thing at a time).

```diff
diff --git a/pkgs/development/libraries/cracklib/default.nix b/pkgs/development/libraries/cracklib/default.nix
index ba5d96a95182..defdb2965173 100644
--- a/pkgs/development/libraries/cracklib/default.nix
+++ b/pkgs/development/libraries/cracklib/default.nix
@@ -15,10 +15,10 @@ stdenv.mkDerivation rec {
     hash = "sha256-yosEmjwtOyIloejRXWE3mOvHSOOVA4jtomlN5Qe6YCA=";
   };

-  nativeBuildInputs = lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) buildPackages.cracklib;
+  nativeBuildInputs = lib.optional (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) buildPackages.cracklib;
   buildInputs = [ zlib gettext ];

-  postPatch = lib.optionalString (stdenv.hostPlatform == stdenv.buildPlatform) ''
+  postPatch = lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
     chmod +x util/cracklib-format
     patchShebangs util
```

kjeremy replied:
This goes against your comment below about using `buildPackages` in `nativeBuildInputs`.

amjoseph-nixpkgs replied:
> This goes against your comment below about using `buildPackages` in `nativeBuildInputs`.

Indeed!  Nice catch.  If this PR were merged CI would have caught that.

The patch above should, in addition to the changes it makes, also delete the `buildPackages.` before `cracklib`.

Commit:
targetPackages_bintools: give it a top-level entry

In order to avoid setting off the unspliced-dependency check, we
need to make sure that the targetPackages.stdenv.cc.bintools used by
gcc gets spliced.  In order for it to be spliced, it needs a
top-level entry in all-packages.nix.  Therefore, this commit adds one.

This also causes us to stop abusing targetPackages to access the
linker, which is a good thing.

Commit:
stdenv: alphabetize list-of-shame

Commit:
stdenv: improve diagnostics for unspliced-dependency check

amjoseph-nixpkgs commented:
Woo hoo, CI passed!

Now about those rebuilds...

Commit:
Revert "gcc: use stdenv.cc.bintools instead of targetPackages_bintools in buildInputs"

This reverts commit 0899b6b8746790eec9dc49ea4f4a6c3def072e3b.

amjoseph-nixpkgs commented:
Sweet.  CI=ok, only 14 packages rebuilt (due to fixing splicing failures).  Reviews are now useful.

amjoseph-nixpkgs commented:
Fixed merge conflict.

kjeremy commented:
@amjoseph-nixpkgs can we merge this soon?

kjeremy commented:
@amjoseph-nixpkgs What do you think about merging this so that it forces the issue of fixing packages for cross? I would personally like it so that I can hunt down the cross issues in all of my out of tree packages.

kjeremy commented:
@amjoseph-nixpkgs 23.11 branched off a while ago. Can we merge this?

kjeremy commented:
@amjoseph-nixpkgs why close this? I found this useful when trying to figure out why some of my cross pkgs were broken.

katexochen commented:
@kjeremy I think this PR was closed in in the light of https://github.com/NixOS/nixpkgs/issues/50105#issuecomment-1902682831. While not necessary, Adam closed some of their open PRs (which is of cause okay if they like to do so).
